Source/Scene/Primitive.js

```diff
@@ -1394,17 +1414,75 @@ define([
             batchTable.setBatchedAttribute(i, center3DHighIndex, encodedCenter.high);
             batchTable.setBatchedAttribute(i, center3DLowIndex, encodedCenter.low);
 
-            var cartographic = ellipsoid.cartesianToCartographic(center, scratchBoundingSphereCartographic);
-            var center2D = projection.project(cartographic, scratchBoundingSphereCenter2D);
-            encodedCenter = EncodedCartesian3.fromCartesian(center2D, scratchBoundingSphereCenterEncoded);
-            batchTable.setBatchedAttribute(i, center2DHighIndex, encodedCenter.high);
-            batchTable.setBatchedAttribute(i, center2DLowIndex, encodedCenter.low);
+            if (!frameState.scene3DOnly) {
+                var cartographic = ellipsoid.cartesianToCartographic(center, scratchBoundingSphereCartographic);
+                var center2D = projection.project(cartographic, scratchBoundingSphereCenter2D);
+                encodedCenter = EncodedCartesian3.fromCartesian(center2D, scratchBoundingSphereCenterEncoded);
+                batchTable.setBatchedAttribute(i, center2DHighIndex, encodedCenter.high);
+                batchTable.setBatchedAttribute(i, center2DLowIndex, encodedCenter.low);
+            }
+
             batchTable.setBatchedAttribute(i, radiusIndex, radius);
         }
 
         primitive._batchTableBoundingSpheresUpdated = true;
     }
 
+    var offsetScratchCartesian = new Cartesian3();
+    function updateBatchTableOffsets(primitive, frameState) {
+        var hasOffset = defined(primitive._batchTableAttributeIndices.offset);
+        if (!hasOffset|| primitive._batchTableOffsetsUpdated || frameState.scene3DOnly) {
+            return;
+        }
+
+        var index2D = primitive._batchTableOffsetAttribute2DIndex;
+
+        var projection = frameState.mapProjection;
+        var ellipsoid = projection.ellipsoid;
+
+        var batchTable = primitive._batchTable;
+        var boundingSpheres = primitive._instanceBoundingSpheres;
+        var length = boundingSpheres.length;
+
+        for (var i = 0; i < length; ++i) {
+            var boundingSphere = boundingSpheres[i];
+            if (!defined(boundingSphere)) {
+                continue;
+            }
+            var offset = batchTable.getBatchedAttribute(i, primitive._batchTableAttributeIndices.offset);
+            if (Cartesian3.equals(offset, Cartesian3.ZERO)) {
+                batchTable.setBatchedAttribute(i, index2D, Cartesian3.ZERO);
+                continue;
+            }
+
+            var modelMatrix = primitive.modelMatrix;
+            if (defined(modelMatrix)) {
+                boundingSphere = BoundingSphere.transform(boundingSphere, modelMatrix, scratchBoundingSphere);
+            }
+
+            var center = boundingSphere.center;
+            center = ellipsoid.scaleToGeodeticSurface(center, center);
```

Comment:
You don't want to modify the bounding sphere right? Should this use a scratch variable?